### ISSUE TITLE
Sentral parameter management functions

### ISSUE THREAD

**Issue by bvernham** [simondlevy/USFS#11]
First, this is great work.

Is is possible to move the Sentral parameter management functions in the library so they are formalized instead of keeping them in the sketch? This would seem to be something that would allow proper update/version control of them.

Thanks

Bruce

**Comment by simondlevy** [comment]
Thanks, Bruce; however, I'm not sure what you mean by moving the parameter management functions in the library.  If you want to set up a code example, that might help.

**Comment by bvernham** [comment]
I would think the following functions would either be in the Sentral library of in their own library and not the sketch itself:

For example:
static void EM7180_set_integer_param (uint8_t param, uint32_t param_val)
static void EM7180_set_WS_params()
static void EM7180_acc_cal_upload()
static void readSenParams()
static void writeSenParams()
static void Accel_cal_check(int16_t accelCount[3])
static void readAccelCal()
static void writeAccCal()
void M24512DFMwriteByte(uint8_t device_address, uint8_t data_address1, uint8_t data_address2, uint8_t  
static void M24512DFMwriteBytes(uint8_t device_address, uint8_t data_address1, uint8_t data_address2, 
static void M24512DFMreadBytes(uint8_t device_address, uint8_t data_address1, uint8_t data_address2, 
static void I2Cscan() 
static void sensorError(const char * errmsg)
static void readParams(uint8_t paramId, uint8_t param[4])

I would think this would be more appropriate to be part of the library itself.

Tanks

Bruce

**Comment by bvernham** [comment]
Then the i2cscan and the M24512DF functions should be part of an additional libraries I would think.

Thanks

**Comment by bvernham** [comment]
Interesting quote from the data sheet:
> A significant number of parameters are used in the SENtral algorithm as it executes, and
these parameters are refined as the SENtral device is used. These include parameters
associated with SENtral’s continuous background calibration function and gyro bias
correction. When SENtral is powered down or otherwise re-initialized, these parameters
also are re-initialized and the parameter refinement process must start over. 

So basically the calibration process is not one and done should be continuous?

> To effectively enable a warm-start process, it is necessary to periodically save all 35
WarmStart parameters, and to reload all of them after SENtral is re-initialized. 

I do not see anywhere in the Sentral data sheet which says the calibration parameters are uploaded automatically, only the SENtral Configuration File..

So I believe the storage and retrieval procedure for the calibration parameters will need to be utilized continuously.

Thanks

Bruce

**Comment by kriswiner** [comment]
The Sentral makes provision for storing 35 parameters which hould allow the user to restart on subsequent power up wherever the Sentral left off in its continuing evolution of its algorithm parameters. These do no include accelerometer calibration parameters or some other magnetometer calibration parameters that are stored in separate locations on the EEPROM. if all of this sounds complicated, it is, made so by the very limited available RAM space for custom programming.

But in general, the USFS board allows 1) storage of user calibration data for mag and accel and 2) storage of warm start parameters used by the SENtral fusion algorithm. Neither are automatic in the sense that the user must  purposely store and retrieve either or both. But the sketch or program can be configured to automatically do both such that on each power on the SENtral starts off with the best available calibration and fusion aprameters for continued high-accuracy absolute orientation estimation.

The caveat is, the calibrations only hold if the temperature and magnetic fild environment are substantially similar each time. Otherwise, recalibration is advisable. As is periodic recalibration from time to time in any case.

So to boil all of this down and repeat, it is possible (if not straightforward) to store calibration data on the EEPROM such that on subsequent power on the absolute orientation estimation starts with and maintains 2-degree heading accuracy.

**Comment by bvernham** [comment]
As the EM7180 SENtral is presently configured, does it report measured or linear acceleration?

Kris's wiki show the calculation of linear acceleration but this data sheet/wiki shows that AX/AY/AZ report linear acceleration.

Which is correct?

Thanks

Bruce

**Comment by kriswiner** [comment]
Not sure what you mean here. One of the standard outputs of the EM7180 is
scaled (called linear in the EM7180 data sheet) acceleration in gs or
meters square second. We have taken advantage of the extra RAM to allow
output of scaled acceleration with gravity removed, but I don't think this
is part of the normal firmware loaded onto the USFS. So instead this is
done on the host by constructing elements of the rotation matrix in a
rather straightforward way.

**Comment by bvernham** [comment]
Is the USFS acceleration with or without gravity removed?

**Comment by kriswiner** [comment]
Standard scaled output has gravity included.

**Comment by bvernham** [comment]
Ok, so I fired up the Teensy and ran the master test.

Everything looks OK.  The accelerators point toward gravity and the yaw points toward northish.

So then I ran the WarmStartAccCal and selected no WS or accel cals.

I then tried tried the figure 8 for a while and then tried to see how well the heading is estimated but the yaw seems to responds very little to rotation around the Z axis.

Am I missing something?

Thanks

Bruce

**Comment by kriswiner** [comment]
Probably.

Which sketch are you using?

Which yaw are you looking at?

How did you verify sensor calibration?

This
<https://github.com/kriswiner/EM7180_SENtral_sensor_hub/wiki/K.-Limits-of-Absolute-Heading-Accuracy-Using-Inexpensive-MEMS-Sensors>
might be helpful as a guide to what kind of perfromance might be expected.

**Comment by bvernham** [comment]
I am using WarmStartAccCal example in Simon Levy's library.  The MasterTest example worked as expected.  I did not have access to cal yet but the output seemed normal. The WarmStartAccCal only has one rotational output "hardware yaw" which shows very little change when rotating the USFS board with the Zaxis pointing down.

**Comment by kriswiner** [comment]
Not sure what the problem is, do the sensor data respond correclty?

" shows very little change when rotating the USFS board with the Zaxis
pointing down" isn't much to go on.

**Comment by kriswiner** [comment]
Couple of things.

This is Simon's decomposition of our sketches reposited here
<https://github.com/kriswiner/EM7180_SENtral_sensor_hub>. For specific
questions about how these sketches perform (or don;t), you will have to ask
him.

If I understand you correctly, the master test sketch works as expected (so
I would use this) and the warm start calibration sketch does not. As far as
I can tell, the warm start calibration sketch isn't intended to produce
accurate orientation estimation. Or at least I see several problems with it
including fixed magnetic declination of 13.8 degrees (is this true for your
location?) and an error is the quaternion divisor:

em7180.setQRateDivisor(0x02); // 100 Hz is really sampling the quats at
200/(1 + 2 ) or 66 Hz, not sure if this is what you intended.

Maybe lots of others (is the yaw constructon correct? What happens if no
warm start parameters have been saved? What happens if no accel calibration
parameters have been saved?).

Bottom line, I don;t know what one should expect running this sketch as I
have never run it. Maybe Simon can respond to your observations.

**Comment by kriswiner** [comment]
Last thing, this sketch makes use of I2C reads/write of ? 32 bytes, the
standard limit for Arduino. I can;t remember if Teensy supports > 32 byte
reads/writes, but the warm start parameters are burst read 128 bytes at a
time, so often the problem is partial read gives garbage answers. We have
seen this on the ESP32/8266 and elsewhere. is tis  problem for the Teensy?
depends on the i2c.h API I would think.

M24512DFMwriteBytes(M24512DFM_DATA_ADDRESS, 0x7f, 0x00, 128, &data[0]); //
Page 254

on line 304 writes 128 bytes at a time  for example.

**Comment by bvernham** [comment]
According to Simon that is the sketch to be used for the purpose of aquiring and storing the WS and accel Cal's based on the page within the Wiki you wrote.

Tomorrow I will try the examples within your GitHub page.

**Comment by kriswiner** [comment]
Hi Greg, Bruce seems to be having trouble using warm start with an MPU9250
USFS and a Teensy. Can you point him to the proper sketch, if any, to try
out?

**Comment by kriswiner** [comment]
Hi Bruce,

 

This sketch and included files was written to support accel cal and warm start for Teensy 3.1/3.2. Give it a try and see if it helps…

 

Best,

Greg

**Comment by bvernham** [comment]
OK the good (and bad news).  I ran the warm start calibration cal from Kris's original Git Hub.

So the mag seems to cal and true north and yaw (after I adjust based on local declination) seems good.

So I apply the warm start calibrations, restart and then load the warm start settings but do not load the accelerate cals.

So then the bad part happens.  For some reason the MPU9250 update rates become 0 Hz and the outputs freeze.

> Scanning...
I2C device found at address 0x28  !
done

EM7180 ROM Version: 0xE69
Should be: 0xE609
EM7180 RAM Version: 0x17435
EM7180 ProductID: 0x80 Should be: 0x80
EM7180 RevisionID: 0x2 Should be: 0x02
A barometer is installed
A temperature sensor is installed
EEPROM detected on the sensor bus!
EEPROM uploaded config file!
EEPROM upload successful!
Send '1' for Warm Start, '0' for no Warm Start
!!!Warm Start active!!!
Send '2' to apply Accelerometer Cal, '0' to not apply Accelerometer Cal
***No Accel Cal***
Beginning Parameter Adjustments
Magnetometer Default Full Scale Range: +/-1000uT
Accelerometer Default Full Scale Range: +/-8g
Gyroscope Default Full Scale Range: +/-2000dps
Magnetometer New Full Scale Range: +/-1000uT
Accelerometer New Full Scale Range: +/-8g
Gyroscope New Full Scale Range: +/-2000dps
 EM7180 run status = normal mode
 EM7180 magnetic anomaly detected
 EM7180 new quaternion result
 EM7180 new mag result
 EM7180 new accel result
 EM7180 new gyro result
 EM7180 sensor status = 0
Actual MagRate = 0 Hz
Actual AccelRate = 0 Hz
Actual GyroRate = 0 Hz
Actual BaroRate = 0 Hz

If I do not load the warm start settings and bypass the accel cals, everything runs and normal.

So the USFS looks good but the sketch "EM71280_MPU9250_BMP280_M24512DFC_WS_Acc_Cal.ino seems strange.

Any ideas how to fix?

Thanks

Bruce

**Comment by kriswiner** [comment]
Sounds like an I2C issue, did you try Greg's suggested "known good" sketch?

**Comment by bvernham** [comment]
I did not get a specific suggestion.  I did not get a link.  I used the sketch from your GitHub "EM71280_MPU9250_BMP280_M24512DFC_WS_Acc_Cal.ino"

**Comment by kriswiner** [comment]
OK, what is your email ? I will forward you the e-mail.

**Comment by bvernham** [comment]
bvernham@juno.com

**Comment by bvernham** [comment]
https://www.sparkfun.com/news/2851?utm_campaign=January%2011%2C%202018&utm_medium=email&_hsenc=p2ANqtz-8IgY8RMl3BThkGpUdp1KRxN-l9fOpag6ZiJWNfCGAP9ZbOxZwRvEFoIv4BZtNoQ_XTK-X39v8XuU41hab4P6PTGPJp7A&_hsmi=68913974&utm_content=68913717&utm_source=hs_email&hsCtaTracking=b0b6118c-c6ac-4407-9c59-4c0c17711ffe%7C4a03f426-3765-44af-9206-492e8d574924 Have you heard about his? ---------- Original Message ----------

**Comment by bvernham** [comment]
Well the new sketch got everything working but I just heard that the MPU9250 is EOL:
![image](https://user-images.githubusercontent.com/5819349/51034616-5b61be00-1575-11e9-9873-f6bb089bc56d.png)

What is next?

**Comment by kriswiner** [comment]
This
<https://www.tindie.com/products/onehorse/all-st-motion-sensor-breakout-board/>
or this
<https://www.tindie.com/products/onehorse/ultimate-sensor-fusion-solution-lsm6dsm-lis2md/>
is to follow. See this
<https://github.com/kriswiner/EM7180_SENtral_sensor_hub/wiki/K.-Limits-of-Absolute-Heading-Accuracy-Using-Inexpensive-MEMS-Sensors>
also.

**Comment by enissimsek** [comment]
So, I just received a few new USFS units from Kris and attempted to use Simon's version of calibration code from "UsfsWarmStartAndAccelCal.cpp" while following instructions at "F. Magnetometer and Accelerometer Calibration" and I see the same problem:

**"The WarmStartAccCal only has one rotational output "hardware yaw" which shows very little change when rotating the USFS board with the Zaxis pointing down."**

This is because the "Hardware yaw" output is changing only with the rotation of the USFS board around Y-axis (the shorter dimension of the board) which I'd expect to be the Pitch. Axis seem to have been switched.

Original EM71280_MPU9250_BMP280_M24512DFC_WS_Acc_Cal.ino works fine.
I can't explain it.

**Comment by kriswiner** [comment]
Is it possible you are mixing LSM6DSM and MPU9250 sketches?

**Comment by simondlevy** [comment]
I'm overseas right now but should be able to work on this next when I get back next week.  For the time being I suggest sticking with Kris's original calibration sketches.

**Comment by enissimsek** [comment]
> Is it possible you are mixing LSM6DSM and MPU9250 sketches?

Are there LSM6DSM- and MPU9250-specific versions of WarmStart/Calibration sketches?
My boards are LSM6DSM and I used Simon's code as is and EM71280_MPU9250_BMP280_M24512DFC_WS_Acc_Cal.ino from your repo.

**Comment by simondlevy** [comment]
Any updates on this issue?  I'm back in my office and have both the MPU9250 and LSM6DSM versions of the board, so I can try and replicate the results if I know what to look for in the output.

**Comment by enissimsek** [comment]
This is where I left it:

Using **USFS LSM6DSM version** when I run **~simondlevy/USFS/examples/WarmStartAndAccelCal/WarmStartAndAccelCal.ino** I get the following output when I rotate the USFS around its Z axis (basically rotate 360 degrees when its on the desk on a breadboard) - I expect the Yaw to change, but it doesn't (seems to be the same issue OP experienced):

USFS ROM Version: 0xE609 (should be: 0xE609)
USFS RAM Version: 0x44579
USFS ProductID: 0x80 Should be: 0x80
USFS RevisionID: 0x2 Should be: 0x02
EEPROM detected on the sensor bus!
EEPROM upload successful!
Send '1' for Warm Start, '0' for no Warm Start **[0]**
***No Warm Start***
Send '2' to apply Accelerometer Cal, '0' to not apply Accelerometer Cal **[0]**
***No Accel Cal***
Beginning Parameter Adjustments
Magnetometer Default Full Scale Range: +/-1000uT
Accelerometer Default Full Scale Range: +/-4g
Gyroscope Default Full Scale Range: +/-2000dps
Magnetometer New Full Scale Range: +/-1000uT
Accelerometer New Full Scale Range: +/-8g
Gyroscope New Full Scale Range: +/-2000dps
 USFS magnetic anomaly detected
ax = 23.42 ay = -8.30 az = 1039.44 mg
Hardware yaw: **13.30**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 23.42 ay = -9.76 az = 1040.90 mg
Hardware yaw: **13.31**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 23.91 ay = -8.30 az = 1039.93 mg
Hardware yaw: **13.31**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 23.91 ay = -8.78 az = 1040.42 mg
Hardware yaw: **13.32**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 23.42 ay = -8.78 az = 1039.93 mg
Hardware yaw: **13.32**


Instead I can only get the Yaw change if I rotate the USFS around its Y-axis (around the shorter dimension of the board):

Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 46.36 ay = 10.74 az = 1037.00 mg
Hardware yaw: **14.28**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 47.34 ay = 10.25 az = 1037.49 mg
Hardware yaw: **14.35**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 46.36 ay = 9.27 az = 1037.49 mg
Hardware yaw: **14.32**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 46.85 ay = 7.32 az = 1037.49 mg
Hardware yaw: **14.14**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 59.54 ay = -134.20 az = 1028.22 mg
Hardware yaw: **5.75**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 57.10 ay = -367.95 az = 959.90 mg
Hardware yaw: **352.92**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 53.68 ay = -467.99 az = 914.51 mg
Hardware yaw: **347.03**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 51.73 ay = -524.60 az = 881.82 mg
Hardware yaw: **343.35**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 47.34 ay = -622.20 az = 809.10 mg
Hardware yaw: **336.66**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 44.41 ay = -694.91 az = 747.13 mg
Hardware yaw: **331.42**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 40.50 ay = -767.14 az = 670.51 mg
Hardware yaw: **325.12**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 40.02 ay = -904.26 az = 440.66 mg
Hardware yaw: **312.56**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 33.18 ay = -958.92 az = 286.94 mg
Hardware yaw: **307.11**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 32.70 ay = -984.30 az = 148.35 mg
Hardware yaw: **300.25**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 25.86 ay = -993.08 az = 26.35 mg
Hardware yaw: **293.42**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 24.40 ay = -993.08 az = -11.71 mg
Hardware yaw: **292.72**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 23.91 ay = -992.10 az = -24.40 mg
Hardware yaw: **290.83**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 30.26 ay = -974.05 az = 205.94 mg
Hardware yaw: **301.99**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 34.65 ay = -891.09 az = 466.04 mg
Hardware yaw: **315.64**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 48.31 ay = -635.38 az = 805.69 mg
Hardware yaw: **335.46**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 60.51 ay = -311.34 az = 989.18 mg
Hardware yaw: **352.79**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 60.02 ay = -168.36 az = 1027.24 mg
Hardware yaw: **0.14**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 60.02 ay = -49.29 az = 1038.95 mg
Hardware yaw: **6.26**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 48.80 ay = 46.85 az = 1035.05 mg
Hardware yaw: **10.94**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 50.75 ay = 75.15 az = 1039.93 mg
Hardware yaw: **13.35**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 54.17 ay = 237.17 az = 1010.16 mg
Hardware yaw: **21.82**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 55.14 ay = 290.36 az = 1000.40 mg
Hardware yaw: **25.33**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 56.61 ay = 401.14 az = 961.36 mg
Hardware yaw: **31.64**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 55.63 ay = 516.30 az = 902.80 mg
Hardware yaw: **38.91**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 52.70 ay = 618.30 az = 835.94 mg
Hardware yaw: **46.17**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 51.73 ay = 768.60 az = 693.94 mg
Hardware yaw: **57.04**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 50.26 ay = 876.94 az = 543.63 mg
Hardware yaw: **67.17**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 51.24 ay = 891.58 az = 519.23 mg
Hardware yaw: **69.54**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 54.66 ay = 979.42 az = 303.54 mg
Hardware yaw: **80.84**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 52.70 ay = 1005.28 az = 185.44 mg
Hardware yaw: **86.09**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 53.68 ay = 1011.62 az = 121.02 mg
Hardware yaw: **90.04**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 52.22 ay = 1018.46 az = 47.34 mg
Hardware yaw: **93.55**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 59.54 ay = 802.27 az = 635.38 mg
Hardware yaw: **57.03**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 49.29 ay = 15.13 az = 1060.42 mg
Hardware yaw: **12.19**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 92.72 ay = 3.90 az = 1035.54 mg
Hardware yaw: **11.88**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 91.26 ay = 2.93 az = 1037.49 mg
Hardware yaw: **12.79**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 78.57 ay = 11.22 az = 1038.95 mg
Hardware yaw: **13.27**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 87.84 ay = 1.95 az = 1037.98 mg
Hardware yaw: **13.44**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 87.84 ay = 1.46 az = 1037.98 mg
Hardware yaw: **13.56**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 87.84 ay = 1.46 az = 1037.49 mg
Hardware yaw: **13.68**
Send '1' to store Warm Start configuration
Send '2' to store Accel Cal
ax = 87.84 ay = 1.46 az = 1037.49 mg
Hardware yaw: **13.75**

**Comment by simondlevy** [comment]
Thanks for your patience.  This yaw bug appears to have been a simple typo that I made in the quaternion => Euler conversion formula, not anything to do with MPU9250 vs LSM6DSM flavors of the USFS.   Once you've verified that my update works, feel free to close the issue or post again if it still appears broken.

**Comment by enissimsek** [comment]
Thanks for the fix - and **verified**.
The issue was started by @bvernham so I'm unable to close.
